File: backend/model/google_book.go
package model

// GoogleBook Google Books APIからのレスポンスモデル
type GoogleBook struct {
	ID           string   `json:"id" example:"zyTCAlFPjgYC"`
	Title        string   `json:"title" example:"Go言語による並行処理"`
	Authors      []string `json:"authors" example:"Katherine Cox-Buday"`
	Description  string   `json:"description" example:"Go言語の並行処理について解説した書籍"`
	ISBN         string   `json:"isbn" example:"9784873118468"`
	ImageURL     string   `json:"image_url" example:"http://books.google.com/books/content?id=..."`
	PublishedDate string   `json:"published_date" example:"2018-06-15"`
}

// GoogleBookSearchRequest 検索リクエスト
type GoogleBookSearchRequest struct {
	Query string `json:"query" validate:"required" example:"Go言語"`
	MaxResults int `json:"max_results" example:"10"`
}

// GoogleBookSearchResponse 検索結果レスポンス
type GoogleBookSearchResponse struct {
	TotalItems int          `json:"total_items" example:"42"`
	Items      []GoogleBook `json:"items"`
}

// ToBookRequest GoogleBookからBookRequestへの変換メソッド
func (gb *GoogleBook) ToBookRequest() BookRequest {
	author := ""
	if len(gb.Authors) > 0 {
		author = gb.Authors[0]
	}
	
	return BookRequest{
		Title:        gb.Title,
		Author:       author,
		Description:  gb.Description,
		ISBN:         gb.ISBN,
		ImageURL:     gb.ImageURL,
		PublishedDate: gb.PublishedDate,
	}
}

File: frontend/src/types/models/googleBook.ts
/**
 * Google Books APIから取得する書籍の型定義
 */

// 業界識別子（ISBN等）の型定義
export interface IndustryIdentifier {
  type: string;
  identifier: string;
}

// バックエンドのモデルに合わせた型定義
export type GoogleBookVolume = {
  id: string;
  title: string;
  authors?: string[];
  description?: string;
  isbn?: string;
  image_url?: string;
  published_date?: string;
};

// バックエンドのレスポンス型に合わせる
export type GoogleBookSearchResponse = {
  items: GoogleBookVolume[];
  total_items: number; // または totalItems
};

export type GoogleBookState = {
  searchQuery: string;
  searchResults: GoogleBookVolume[];
  selectedBook: GoogleBookVolume | null;
  updateSearchQuery: (query: string) => void;
  setSearchResults: (results: GoogleBookVolume[]) => void;
  selectBook: (book: GoogleBookVolume | null) => void;
  resetGoogleBookState: () => void;
};
